### DISCUSSION TITLE
Charon crashes when executing "swanctl --load-conns" with miedated_by is used

### DISCUSSION THREAD

**Discussion by bluven** [strongswan/strongswan#1569] [Q&A] [answered]
I was trying out the mediation plugin, but everytime I executed "swanctl --list-conns", I got error like this:
```
/ # swanctl --load-conns
loaded connection 'mediator'
load-conn request failed: No error information
```
soon I found that the charon crashed:
```
Starting strongSwan 5.9.9 IPsec [starter]...
00[DMN] Starting IKE charon daemon (strongSwan 5.9.9, Linux 5.15.0-60-generic, x86_64)
00[LIB] providers loaded by OpenSSL: legacy default
00[CFG] install DNS servers in '/etc/resolv.conf'
00[CFG] loading ca certificates from '/etc/ipsec.d/cacerts'
00[CFG]   loaded ca certificate "C=CH, O=strongSwan, CN=strongSwan Root CA" from '/etc/ipsec.d/cacerts/ca.crt'
00[CFG] loading aa certificates from '/etc/ipsec.d/aacerts'
00[CFG] loading ocsp signer certificates from '/etc/ipsec.d/ocspcerts'
00[CFG] loading attribute certificates from '/etc/ipsec.d/acerts'
00[CFG] loading crls from '/etc/ipsec.d/crls'
00[CFG] loading secrets from '/etc/ipsec.secrets'
00[CFG]   loaded ED25519 private key from '/etc/ipsec.d/private/edge2.key'
00[CFG] loaded 0 RADIUS server configurations
00[LIB] loaded plugins: charon aesni sha3 md4 mgf1 random nonce x509 revocation constraints pubkey pkcs1 pkcs7 pkcs12 pgp dnskey sshkey pem openssl pkcs8 fips-prf gmp curve25519 chapoly xcbc cmac kdf gcm ntru drbg newhope curl files sqlite attr kernel-netlink resolve socket-default farp stroke vici updown eap-identity eap-sim eap-aka eap-aka-3gpp2 eap-simaka-pseudonym eap-simaka-reauth eap-md5 eap-mschapv2 eap-dynamic eap-radius eap-tls xauth-generic xauth-eap dhcp unity counters
charon (10) started after 20 ms
sh: addr2line: not found
 dumping 1 stack frame addresses:
  /lib/ld-musl-x86_64.so.1 @ 0x00007f644a5b7000 [0x00007f644a600acd]
sh: addr2line: not found
    -> 
charon has died -- restart scheduled (5sec)
00[DMN] Starting IKE charon daemon (strongSwan 5.9.9, Linux 5.15.0-60-generic, x86_64)
00[LIB] providers loaded by OpenSSL: legacy default
00[CFG] install DNS servers in '/etc/resolv.conf'
00[KNL] unable to create IPv4 routing table rule
00[KNL] unable to create IPv6 routing table rule
00[CFG] loading ca certificates from '/etc/ipsec.d/cacerts'
00[CFG]   loaded ca certificate "C=CH, O=strongSwan, CN=strongSwan Root CA" from '/etc/ipsec.d/cacerts/ca.crt'
00[CFG] loading aa certificates from '/etc/ipsec.d/aacerts'
00[CFG] loading ocsp signer certificates from '/etc/ipsec.d/ocspcerts'
00[CFG] loading attribute certificates from '/etc/ipsec.d/acerts'
00[CFG] loading crls from '/etc/ipsec.d/crls'
00[CFG] loading secrets from '/etc/ipsec.secrets'
00[CFG]   loaded ED25519 private key from '/etc/ipsec.d/private/edge2.key'
00[CFG] loaded 0 RADIUS server configurations
00[LIB] loaded plugins: charon aesni sha3 md4 mgf1 random nonce x509 revocation constraints pubkey pkcs1 pkcs7 pkcs12 pgp dnskey sshkey pem openssl pkcs8 fips-prf gmp curve25519 chapoly xcbc cmac kdf gcm ntru drbg newhope curl files sqlite attr kernel-netlink resolve socket-default farp stroke vici updown eap-identity eap-sim eap-aka eap-aka-3gpp2 eap-simaka-pseudonym eap-simaka-reauth eap-md5 eap-mschapv2 eap-dynamic eap-radius eap-tls xauth-generic xauth-eap dhcp unity counters
charon (41) started after 20 ms
```

I compiled strongswan 5.9.9 myself using docker, here is the Dockerfile:
```
#
# StrongSwan VPN + Alpine Linux
#

FROM alpine:edge

ENV STRONGSWAN_RELEASE https://download.strongswan.org/strongswan.tar.bz2

RUN sed -i 's/dl-cdn.alpinelinux.org/mirrors.tuna.tsinghua.edu.cn/g' /etc/apk/repositories && \
    apk --update add build-base \
            ca-certificates \
            curl \
            curl-dev \
            gmp-dev \
            sqlite-dev \
            libunwind-dev \
            ip6tables \
            iproute2 \
            iptables-dev \
            openssl \
            openssl-dev && \
    mkdir -p /tmp/strongswan && \
    curl -Lo /tmp/strongswan.tar.bz2 $STRONGSWAN_RELEASE && \
    tar --strip-components=1 -C /tmp/strongswan -xjf /tmp/strongswan.tar.bz2 && \
    cd /tmp/strongswan && \
    ./configure --prefix=/usr \
            --sysconfdir=/etc \
            --libexecdir=/usr/lib \
            --with-ipsecdir=/usr/lib/strongswan \
            --enable-aesni \
            --enable-chapoly \
            --enable-cmd \
            --enable-curl \
            --enable-dhcp \
            --enable-eap-dynamic \
            --enable-eap-identity \
            --enable-eap-md5 \
            --enable-eap-mschapv2 \
            --enable-eap-radius \
            --enable-eap-tls \
            --enable-farp \
            --enable-files \
            --enable-gcm \
            --enable-md4 \
            --enable-newhope \
            --enable-ntru \
            --enable-openssl \
            --enable-sha3 \
            --enable-shared \
            --enable-gmp \
            --enable-eap-aka \
            --enable-eap-aka-3gpp2 \
            --enable-eap-sim \
            --enable-eap-simaka-pseudonym \
            --enable-eap-simaka-reauth \
            --enable-sqlite \
            --enable-unity \
            --enable-xauth-eap \
            --enable-xauth-generic \
            --enable-mediation \
            --enable-unwind-backtraces \
            --disable-aes \
            --disable-des \
            --disable-hmac \
            --disable-ikev1 \
            --disable-md5 \
            --disable-rc2 \
            --disable-sha1 \
            --disable-sha2 \
            --disable-static && \
    make && \
    make install && \
    rm -rf /tmp/* && \
    apk del build-base curl-dev openssl-dev && \
    rm -rf /var/cache/apk/*

EXPOSE 500/udp \
       4500/udp

ENTRYPOINT ["/usr/sbin/ipsec"]
CMD ["start", "--nofork"]
```

I didnt' change swanctl.conf/ipsec.conf since I put all connections under "/etc/swanctl/conf.d/connections.conf" , that file looks like this:
```
connections {
    mediator {
        version = 2
        proposals = default
        local_addrs = 192.168.0.3
        remote_addrs = 10.22.46.20
        mediation = yes

        local {
            auth = pubkey
            id = C=CH, O=strongswan, CN=edge2.strongswan.org
            certs = /etc/ipsec.d/certs/edge2.crt
        }

        remote {
	        auth = pubkey
	        id = C=CH, O=strongswan, CN=connector.strongswan.org
        }
    }

    edge1 {
		version = 2
		proposals = default
        	local_addrs = 192.168.0.3
		remote_addrs = 172.17.0.2
		mediated_by = mediator

		local {
			auth = pubkey
			id = C=CH, O=strongswan, CN=edge2.strongswan.org
                       certs = /etc/ipsec.d/certs/edge2.crt
		}

		remote {
			auth = pubkey
			id = C=CH, O=strongswan, CN=edge1.strongswan.org
		}

		children {
			edge1 {
				remote_ts = 172.17.0.0/16
				local_ts = 192.168.0.0/16
				esp_proposals = default
			}
		}
	}

    connector {
		version = 2
		proposals = default
                local_addrs = 192.168.0.3
                remote_addrs = 10.22.46.20

		local {
			auth = pubkey
			id = C=CH, O=strongswan, CN=edge2.strongswan.org
                       certs = /etc/ipsec.d/certs/edge2.crt
		}

		remote {
			auth = pubkey
			id = C=CH, O=strongswan, CN=connector.strongswan.org
		}

		children {
			connector {
				remote_ts = 2.2.0.0/16
				local_ts = 192.168.0.0/16
				esp_proposals = default
			}
		}
	}
}
```
I think the edge1 connection caused the crash because when comment the "mediated_by = mediator" line, "swanctl --load-conns" could succeed.

I run strongswan like this:
```
#!/bin/bash
docker run -d \
    --name strongswan \
    --volume /lib/modules:/lib/modules:ro \
    -v $PWD/keypairs/ca.crt:/etc/ipsec.d/cacerts/ca.crt \
    -v $PWD/keypairs/edge2.crt:/etc/ipsec.d/certs/edge2.crt \
    -v $PWD/keypairs/edge2.key:/etc/ipsec.d/private/edge2.key \
    -v $PWD/keypairs/ipsec.secrets:/etc/ipsec.secrets \
    -v $PWD/connections.conf:/etc/swanctl/conf.d/connections.conf \
    --privileged \
    strongswan:5.9.9
```

I have googled for a long time but found nothing similar like my problem, I also tried another verion 5.9.4, but both 5.9.9 and 5.9.4 crash. I wanted to use bfd_backtraces to figure out what happened, but failed to make compilation pass so I used unwind_backtraces, but it seems the backtrace didn't work well.

I have done my best and now is in despair, so any help is welcome and thanks in advance.

**Comment by tobiasbrunner** [comment, marked as answer]
It's a bug caused by 229cdf6bc8f24f9ce445943117c518dc6bb876a4 that's triggered if no `mediation_peer` is configured in the connection that has `mediated_by` set. Unfortunately, we didn't notice because [our test scenarios](https://www.strongswan.org/testing/testresults/p2pnat/) both set `mediation_peer`. If no peer ID is configured, the code tries to use `remote.id` as peer ID, but since the referenced commit, the list doesn't store `auth_cfg_t` objects directly anymore.

I've pushed a fix to the _1569-mediated-peer-fallback_ branch. As a workaround, make sure to configure `mediation_peer` (could be the same identity as `remote.id`).

**Reply by bluven** [reply]
@tobiasbrunner Thanks for your quick reply, I'll give it a try.

**Reply by alichen-xxx** [reply]
Hello, I am using the latest version 5.9.14 and also met the issue

**Reply by alichen-xxx** [reply]
Jan 15 14:21:39 debian12-KVM-gui charon[116031]: 04[LIB]  dumping 16 stack frame addresses:
Jan 15 14:21:39 debian12-KVM-gui charon[116031]: 04[LIB]   /lib/x86_64-linux-gnu/libc.so.6 @ 0x7fb783f78000 [0x7fb783fb4050]
Jan 15 14:21:39 debian12-KVM-gui charon[116031]: 04[LIB]     -> ??:?
Jan 15 14:21:39 debian12-KVM-gui charon[116031]: 04[LIB]   /usr/lib/ipsec/libcharon.so.0 @ 0x7fb78416c000 [0x7fb78419b16b]
Jan 15 14:21:39 debian12-KVM-gui charon[116031]: 04[LIB]     -> /home/ute/strongswan-5.9.14/src/libcharon/sa/child_sa.c:797 (discriminator 4)
Jan 15 14:21:39 debian12-KVM-gui charon[116031]: 04[LIB]   /usr/lib/ipsec/plugins/libstrongswan-connmark.so @ 0x7fb783b20000 [0x7fb783b211cf]
Jan 15 14:21:39 debian12-KVM-gui charon[116031]: 04[LIB]     -> ??:0
Jan 15 14:21:39 debian12-KVM-gui charon[116031]: 04[LIB]   /usr/lib/ipsec/plugins/libstrongswan-connmark.so @ 0x7fb783b20000 [0x7fb783b21dad]
Jan 15 14:21:39 debian12-KVM-gui charon[116031]: 04[LIB]     -> ??:0
Jan 15 14:21:39 debian12-KVM-gui charon[116031]: 04[LIB]   /usr/lib/ipsec/libcharon.so.0 @ 0x7fb78416c000 [0x7fb78417bd89]
Jan 15 14:21:39 debian12-KVM-gui charon[116031]: 04[LIB]     -> /home/ute/strongswan-5.9.14/src/libcharon/bus/bus.c:724
Jan 15 14:21:39 debian12-KVM-gui charon[116031]: 04[LIB]   /usr/lib/ipsec/libcharon.so.0 @ 0x7fb78416c000 [0x7fb7841b6dfb]
Jan 15 14:21:39 debian12-KVM-gui charon[116031]: 04[LIB]     -> /home/ute/strongswan-5.9.14/src/libcharon/sa/ikev2/tasks/child_create.c:1547
Jan 15 14:21:39 debian12-KVM-gui charon[116031]: 04[LIB]   /usr/lib/ipsec/libcharon.so.0 @ 0x7fb78416c000 [0x7fb7841afdbf]
Jan 15 14:21:39 debian12-KVM-gui charon[116031]: 04[LIB]     -> /home/ute/strongswan-5.9.14/src/libcharon/sa/ikev2/task_manager_v2.c:964
Jan 15 14:21:39 debian12-KVM-gui charon[116031]: 04[LIB]   /usr/lib/ipsec/libcharon.so.0 @ 0x7fb78416c000 [0x7fb78419db4f]
Jan 15 14:21:39 debian12-KVM-gui charon[116031]: 04[LIB]     -> /home/ute/strongswan-5.9.14/src/libcharon/sa/ike_sa.c:1637
Jan 15 14:21:39 debian12-KVM-gui charon[116031]: 04[LIB]   /usr/lib/ipsec/libcharon.so.0 @ 0x7fb78416c000 [0x7fb7841acf70]
Jan 15 14:21:39 debian12-KVM-gui charon[116031]: 04[LIB]     -> /home/ute/strongswan-5.9.14/src/libcharon/sa/ikev2/task_manager_v2.c:1471
Jan 15 14:21:39 debian12-KVM-gui charon[116031]: 04[LIB]   /usr/lib/ipsec/libcharon.so.0 @ 0x7fb78416c000 [0x7fb7841af035]
Jan 15 14:21:39 debian12-KVM-gui charon[116031]: 04[LIB]     -> /home/ute/strongswan-5.9.14/src/libcharon/sa/ikev2/task_manager_v2.c:1911
Jan 15 14:21:39 debian12-KVM-gui charon[116031]: 04[LIB]   /usr/lib/ipsec/libcharon.so.0 @ 0x7fb78416c000 [0x7fb78419db4f]
Jan 15 14:21:39 debian12-KVM-gui charon[116031]: 04[LIB]     -> /home/ute/strongswan-5.9.14/src/libcharon/sa/ike_sa.c:1637
Jan 15 14:21:39 debian12-KVM-gui charon[116031]: 04[LIB]   /usr/lib/ipsec/libcharon.so.0 @ 0x7fb78416c000 [0x7fb784196d0d]
Jan 15 14:21:39 debian12-KVM-gui charon[116031]: 04[LIB]     -> /home/ute/strongswan-5.9.14/src/libcharon/processing/jobs/process_message_job.c:75
Jan 15 14:21:39 debian12-KVM-gui charon[116031]: 04[LIB]   /usr/lib/ipsec/libstrongswan.so.0 @ 0x7fb784204000 [0x7fb7842408b6]
Jan 15 14:21:39 debian12-KVM-gui charon[116031]: 04[LIB]     -> /home/ute/strongswan-5.9.14/src/libstrongswan/processing/processor.c:262
Jan 15 14:21:39 debian12-KVM-gui charon[116031]: 04[LIB]   /usr/lib/ipsec/libstrongswan.so.0 @ 0x7fb784204000 [0x7fb784253c40]
Jan 15 14:21:39 debian12-KVM-gui charon[116031]: 04[LIB]     -> /home/ute/strongswan-5.9.14/src/libstrongswan/threading/thread.c:337 (discriminator 4)
Jan 15 14:21:39 debian12-KVM-gui charon[116031]: 04[LIB]   /lib/x86_64-linux-gnu/libc.so.6 @ 0x7fb783f78000 [0x7fb7840011f5]
Jan 15 14:21:39 debian12-KVM-gui charon[116031]: 04[LIB]     -> ??:?
Jan 15 14:21:39 debian12-KVM-gui charon[116031]: 04[LIB]   /lib/x86_64-linux-gnu/libc.so.6 @ 0x7fb783f78000 [0x7fb7840818dc]
Jan 15 14:21:39 debian12-KVM-gui charon[116031]: 04[LIB]     -> ??:?
Jan 15 14:21:39 debian12-KVM-gui ipsec[116031]:   /usr/lib/ipsec/libcharon.so.0 @ 0x7fb78416c000 [0x7fb78419b16b]
Jan 15 14:21:39 debian12-KVM-gui ipsec[116031]:     -> /home/ute/strongswan-5.9.14/src/libcharon/sa/child_sa.c:797 (discriminator 4)
Jan 15 14:21:39 debian12-KVM-gui ipsec[116031]:   /usr/lib/ipsec/libcharon.so.0 @ 0x7fb78416c000 [0x7fb78417bd89]
Jan 15 14:21:39 debian12-KVM-gui ipsec[116031]:     -> /home/ute/strongswan-5.9.14/src/libcharon/bus/bus.c:724
Jan 15 14:21:39 debian12-KVM-gui ipsec[116031]:   /usr/lib/ipsec/libcharon.so.0 @ 0x7fb78416c000 [0x7fb7841b6dfb]
Jan 15 14:21:39 debian12-KVM-gui ipsec[116031]:     -> /home/ute/strongswan-5.9.14/src/libcharon/sa/ikev2/tasks/child_create.c:1547
Jan 15 14:21:39 debian12-KVM-gui ipsec[116031]:   /usr/lib/ipsec/libcharon.so.0 @ 0x7fb78416c000 [0x7fb7841afdbf]
Jan 15 14:21:39 debian12-KVM-gui ipsec[116031]:     -> /home/ute/strongswan-5.9.14/src/libcharon/sa/ikev2/task_manager_v2.c:964
Jan 15 14:21:39 debian12-KVM-gui ipsec[116031]:   /usr/lib/ipsec/libcharon.so.0 @ 0x7fb78416c000 [0x7fb78419db4f]
Jan 15 14:21:39 debian12-KVM-gui ipsec[116031]:     -> /home/ute/strongswan-5.9.14/src/libcharon/sa/ike_sa.c:1637
Jan 15 14:21:39 debian12-KVM-gui ipsec[116031]:   /usr/lib/ipsec/libcharon.so.0 @ 0x7fb78416c000 [0x7fb7841acf70]
Jan 15 14:21:39 debian12-KVM-gui ipsec[116031]:     -> /home/ute/strongswan-5.9.14/src/libcharon/sa/ikev2/task_manager_v2.c:1471
Jan 15 14:21:39 debian12-KVM-gui ipsec[116031]:   /usr/lib/ipsec/libcharon.so.0 @ 0x7fb78416c000 [0x7fb7841af035]
Jan 15 14:21:39 debian12-KVM-gui ipsec[116031]:     -> /home/ute/strongswan-5.9.14/src/libcharon/sa/ikev2/task_manager_v2.c:1911
Jan 15 14:21:39 debian12-KVM-gui ipsec[116031]:   /usr/lib/ipsec/libcharon.so.0 @ 0x7fb78416c000 [0x7fb78419db4f]
Jan 15 14:21:39 debian12-KVM-gui ipsec[116031]:     -> /home/ute/strongswan-5.9.14/src/libcharon/sa/ike_sa.c:1637
Jan 15 14:21:39 debian12-KVM-gui ipsec[116031]:   /usr/lib/ipsec/libcharon.so.0 @ 0x7fb78416c000 [0x7fb784196d0d]
Jan 15 14:21:39 debian12-KVM-gui ipsec[116031]:     -> /home/ute/strongswan-5.9.14/src/libcharon/processing/jobs/process_message_job.c:75
Jan 15 14:21:39 debian12-KVM-gui charon[116031]: 04[DMN] killing ourself, received critical signal
Jan 15 14:21:39 debian12-KVM-gui ipsec[116027]: charon has died -- restart scheduled (5sec)
Jan 15 14:21:39 debian12-KVM-gui ipsec_starter[116027]: charon has died -- restart scheduled (5sec)
Jan 15 14:21:44 debian12-KVM-gui charon[117979]: 00[DMN] Starting IKE charon daemon (strongSwan 5.9.14, Linux 6.1.0-41-amd64, x86_64)
Jan 15 14:21:44 debian12-KVM-gui charon[117979]: 00[LIB] providers loaded by OpenSSL: legacy default
Jan 15 14:21:44 debian12-KVM-gui charon[117979]: 00[CFG] using '/sbin/resolvconf' to install DNS servers
Jan 15 14:21:44 debian12-KVM-gui charon[117979]: 00[KNL] unable to create IPv4 routing table rule
Jan 15 14:21:44 debian12-KVM-gui charon[117979]: 00[KNL] unable to create IPv6 routing table rule
Jan 15 14:21:44 debian12-KVM-gui charon[117979]: 00[CFG] loading ca certificates from '/etc/ipsec.d/cacerts'
Jan 15 14:21:44 debian12-KVM-gui charon[117979]: 00[CFG] loading aa certificates from '/etc/ipsec.d/aacerts'
Jan 15 14:21:44 debian12-KVM-gui charon[117979]: 00[CFG] loading ocsp signer certificates from '/etc/ipsec.d/ocspcerts'
Jan 15 14:21:44 debian12-KVM-gui charon[117979]: 00[CFG] loading attribute certificates from '/etc/ipsec.d/acerts'
Jan 15 14:21:44 debian12-KVM-gui charon[117979]: 00[CFG] loading crls from '/etc/ipsec.d/crls'
Jan 15 14:21:44 debian12-KVM-gui charon[117979]: 00[CFG] loading secrets from '/etc/ipsec.secrets'
Jan 15 14:21:44 debian12-KVM-gui charon[117979]: 00[LIB] loaded plugins: charon aes aesni agent connmark eap-mschapv2 gmp mgf1 des rc2 sha2 sha1 md5 random nonce x509 revocation constraints pubkey pkcs1 pkcs7 pkcs12 pgp dnskey sshkey pem openssl pkcs8 fips-prf curve25519 xcbc cmac hmac kdf gcm drbg attr kernel-netlink resolve socket-default stroke vici updown xauth-generic counters
Jan 15 14:21:44 debian12-KVM-gui charon[117979]: 00[JOB] spawning 16 worker threads

**Reply by tobiasbrunner** [reply]
> Hello, I am using the latest version 5.9.14 and also met the issue

Clearly it's not the latest version and your crash has absolutely nothing to do with the issue discussed here. Your problem is almost certainly because you load an incompatible plugin (`connmark`) from an earlier/different build.

**Reply by alichen-xxx** [reply]
Thanks for your quick comments. And I am testing the ipsec between RU and VM. I install 5.9.14  to match with my RU unit. For load an incompatible plugin , Do you mean there is issue with my installation?

**Reply by alichen-xxx** [reply]
Thanks for your quick support, issue has been resolved after reinstalling the plugin (connmark). Thanks very much.
